### PR TITLE
chore: implement Artifact stream

### PR DESCRIPTION
**Pull request by mxschmitt**

### CODE REVIEW

File: src/Playwright/Core/Stream.cs [hunk 1]
```diff
@@ -42,8 +46,55 @@ internal Stream(IChannelOwner parent, string guid) : base(parent, guid)
...
         public StreamChannel Channel { get; }
 
-        public async ValueTask DisposeAsync() => await Channel.CloseAsync().ConfigureAwait(false);
+        public bool IsClosed { get; set; }
+
+        public StreamImpl StreamImpl => new(this);
+
+        public Task<byte[]> ReadAsync(int size) => Channel.ReadAsync(size);
+
+        public async ValueTask DisposeAsync() => await CloseAsync().ConfigureAwait(false);
```
Comment (pavelfeldman) [inline src/Playwright/Core/Stream.cs:55]: public ValueTask DisposeAsync() => CloseAsync();

File: src/Playwright/Core/Stream.cs [hunk 2]
```diff
@@ -42,8 +46,55 @@ internal Stream(IChannelOwner parent, string guid) : base(parent, guid)
...
+    internal class StreamImpl : System.IO.Stream
+    {
+        private readonly Stream _stream;
+
+        internal StreamImpl(Stream stream)
+        {
+            _stream = stream;
+        }
+
+        public override bool CanRead => !_stream.IsClosed;
```
Comment (pavelfeldman) [inline src/Playwright/Core/Stream.cs:73]: This should be always true - your stream supports reading.

File: src/Playwright/Core/Stream.cs [hunk 3]
```diff
@@ -42,8 +46,55 @@ internal Stream(IChannelOwner parent, string guid) : base(parent, guid)
...
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => throw new NotImplementedException();
```
Comment (pavelfeldman) [inline src/Playwright/Core/Stream.cs:79]: Here and below, throw NotSupportedException instead.

File: src/Playwright/Core/Stream.cs [hunk 4]
```diff
@@ -42,8 +46,55 @@ internal Stream(IChannelOwner parent, string guid) : base(parent, guid)
 
         public StreamChannel Channel { get; }
 
-        public async ValueTask DisposeAsync() => await Channel.CloseAsync().ConfigureAwait(false);
+        public bool IsClosed { get; set; }
```
Comment (pavelfeldman) [inline src/Playwright/Core/Stream.cs:49]: I don't think you need this.

File: src/Playwright/Core/Stream.cs [hunk 5]
```diff
@@ -23,10 +23,14 @@
  */
 
 using System;
+using System.IO;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Playwright.Transport;
 using Microsoft.Playwright.Transport.Channels;
 
+#pragma warning disable SA1402 // A C# code file contains more than one unique type.
```
Comment (pavelfeldman) [inline src/Playwright/Core/Stream.cs:32]: Don't violate, put it into a separate file.

File: src/Playwright/Core/Stream.cs (hunk 1 above)
Comment (mxschmitt) [reply src/Playwright/Core/Stream.cs:55]: I thought the same but ValueTask != Task:

image

We do the same in other places too.

Comment (campersau) [reply src/Playwright/Core/Stream.cs:51]:
```suggestion
        public ValueTask DisposeAsync() => new ValueTask(CloseAsync());
```
This would work.

**Pull request merged**